fccloudapi.php

```diff
@@ -171,56 +171,92 @@ public function getObjectName()
     }
 }
 
+
 /**
- * Base MetadataRecord class
+ * Metadata attribute value utility methods for typecasting.
  * @package codelathe\fccloudapi
  */
-abstract class AbstractMetadataRecord extends DataRecord
+trait MetadataAttributeTypeCasterTrait
 {
-    const TYPE_TEXT = 1;
-    const TYPE_INTEGER = 2;
-    const TYPE_DECIMAL = 3;
-    const TYPE_BOOLEAN = 4;
-    const TYPE_DATE = 5;
-    const TYPE_ENUMERATION = 6;
-    const TYPE_ARRAY = 7;
-
     /**
+     * Cast response data to type
+     *
      * @param mixed $data
      * @param int $type
      * @return array|bool|\DateTime|float|int
      */
-    protected function castToType($data, int $type)
+    public function castToType($data, int $type)
     {
         switch ($type) {
-            case self::TYPE_INTEGER:
-                return (int) $data;
-            case self::TYPE_DECIMAL:
-                return (float) $data;
-            case self::TYPE_BOOLEAN:
+            case AbstractMetadataRecord::TYPE_INTEGER:
+                return strlen($data) ? (int) $data : null;
+            case AbstractMetadataRecord::TYPE_DECIMAL:
+                return strlen($data) ? (float) $data : null;
+            case AbstractMetadataRecord::TYPE_BOOLEAN:
                 return !!json_decode($data);
-            case self::TYPE_DATE:
-                return \DateTime::createFromFormat('Y-m-d H:i:s', $data);
-            case self::TYPE_ARRAY:
-                return explode(',', $data);
+            case AbstractMetadataRecord::TYPE_DATE:
+                $date = \DateTime::createFromFormat('Y-m-d H:i:s', $data);
+                
+                return $date instanceof \DateTime ? $date : null;
+            case AbstractMetadataRecord::TYPE_ARRAY:
+                return strlen($data) ? explode(',', $data) : [];
             default:
                 return $data;
         }
     }
+
+    /**
+     * Best effort reverse cast of attribute value for submission
+     * to api endpoint.
+     *
+     * @param $data
+     * @param int $type
+     * @return string|int|float|bool
+     */
+    protected function reverseCastFromType($data, int $type)
+    {
+        switch ($type) {
+            case AbstractMetadataRecord::TYPE_DATE:
+                return $data instanceof \DateTime ? $data->format('Y-m-d H:i:s') : $data;
+            case AbstractMetadataRecord::TYPE_INTEGER:
+            case AbstractMetadataRecord::TYPE_DECIMAL:
+            case AbstractMetadataRecord::TYPE_BOOLEAN:
+                return json_encode($data);
+            case AbstractMetadataRecord::TYPE_ARRAY:
+                return is_array($data) ? implode(',', $data): $data;
+        }
+
+        return $data;
+    }
 }
 
 /**
- * Class MetadataSetRecord
+ * Hold common constants, methods, etc.
+ * @package codelathe\fccloudapi
+ */
+abstract class AbstractMetadataRecord extends DataRecord
```

Comment:
Is there a need for that abstraction? If I see correctly it's only used as a base class in one place (for another abstract class), so isn't it better to combine both in one?

Reply:
MetadataValueRecord actually also [inherits AbstractMetadataRecord](https://github.com/codelathe/filecloudapi-php/blob/master/fccloudapi.php#L408). It can't be seen in this diff as it was introduced in a prior PR.